### ISSUE TITLE
flink: after rewrite, the two small files are rewritten into the same two small files as before 

### ISSUE THREAD

**Issue by wuwenchi** [apache/iceberg#4074]
Here is a table without partitions with two data files in parquet format: 
- a.parquet 200M
- b.parquet 250M

I started a task of flink's rewriteDataFiles, targetSize is set to 300M.
After waiting for the rewrite to complete, two new files are generated. 
These two new files are exactly the same as the previous two old files.
This situation should not generate new data again. A bit of a waste of performance...

**Comment by RussellSpitzer** [comment]
This is something that will be fixed a little in the latest release, one of our issues is that the algorithm basically could not handle compacting files which were not of the correct size, but when combined would be too large. Now Files are combined at the offset level so more efficient compaction should be possible. Additionally I believe we have other work that goes through and checks to see if any tasks for rewrite are generated that would end up being a noop and skips them.

#3292

**Comment by wuwenchi** [comment]
@RussellSpitzer  Thank you for your answer. I incorporated the pr you mentioned, but the phenomenon persists. 
I followed the process and the problem seems to be in the **isPartialFileScan** function.
This function is fine if using avro format data files.
However, when using data files in parquet format, since the initial state of the parquet file itself has a 4-byte offset, the judgment here is wrong, even if it is a complete parquet file, it will return true here.

Should a judgment be made on the file format here? If it is in parquet format, we need to add an initial offset of 4 bytes to **fileScanTask.length()**.
